File: backend/services/pdfService.js
import { PDFLoader } from "@langchain/community/document_loaders/fs/pdf";
import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";
import { MemoryVectorStore } from "langchain/vectorstores/memory";
import { HuggingFaceTransformersEmbeddings } from "@langchain/community/embeddings/hf_transformers";

// Singleton embeddings model (Xenova/all-MiniLM-L6-v2 — same as before)
export const embeddings = new HuggingFaceTransformersEmbeddings({
  modelName: "Xenova/all-MiniLM-L6-v2",
});

// In-memory vector store (replaces documentStore array + cosineSimilarity)
let vectorStore = null;

export function getVectorStore() {
  return vectorStore;
}

export async function processPDF(req, res) {
  try {
    console.log("File path:", req.file.path);

    // 1. Load PDF — replaces pdf-parse + fs.readFileSync
    const loader = new PDFLoader(req.file.path);
    const docs = await loader.load();

    // 2. Chunk — replaces your chunkText utility
    const splitter = new RecursiveCharacterTextSplitter({
      chunkSize: 500,
      chunkOverlap: 50, // small overlap so context isn't lost at boundaries
    });
    const chunks = await splitter.splitDocuments(docs);

    // 3. Embed + store — replaces your for-loop with createEmbedding + documentStore.push
    vectorStore = await MemoryVectorStore.fromDocuments(chunks, embeddings);

    console.log(`Stored ${chunks.length} chunks in vector store`);
    res.json({ message: "Document processed successfully" });
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "PDF processing failed" });
  }
}

File: backend/services/pdf_service.py
from pathlib import Path
from dotenv import load_dotenv
from langchain_community.document_loaders import PyPDFLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.vectorstores import FAISS
from langchain_community.embeddings import HuggingFaceEmbeddings
from fastapi import UploadFile
from fastapi.responses import JSONResponse
import shutil
import os

load_dotenv(dotenv_path=Path(__file__).parent.parent / ".env")

vector_store = None
_embeddings = None

def get_embeddings():
    global _embeddings
    if _embeddings is None:
        print("Loading embeddings model...")
        _embeddings = HuggingFaceEmbeddings(
            model_name="all-MiniLM-L6-v2",
            model_kwargs={"device": "cpu"},
            encode_kwargs={"normalize_embeddings": True}
        )
        print("Embeddings model loaded")
    return _embeddings

def get_vector_store():
    return vector_store

async def process_pdf(file: UploadFile):
    global vector_store
    try:
        file_path = f"uploads/{file.filename}"
        with open(file_path, "wb") as buffer:
            shutil.copyfileobj(file.file, buffer)

        print(f"File saved: {file_path}")

        loader = PyPDFLoader(file_path)
        docs = loader.load()

        splitter = RecursiveCharacterTextSplitter(
            chunk_size=500,
            chunk_overlap=50,
        )
        chunks = splitter.split_documents(docs)

        print(f"Creating vector store with {len(chunks)} chunks...")
        vector_store = FAISS.from_documents(chunks, get_embeddings())

        print(f"Stored {len(chunks)} chunks in vector store")
        return JSONResponse(content={"message": "Document processed successfully"})

    except Exception as e:
        print(f"PDF processing error: {e}")
        return JSONResponse(status_code=500, content={"error": "PDF processing failed", "detail": str(e)})